### PR TITLE
Story/fl 1161/task/fl 1236 add timebox criteria to form data exporter

### PR DESCRIPTION
**Pull request by samkeen**

### CODE REVIEW

File: src/Tools/FormExporter.php [hunk 1]
```diff
@@ -285,8 +291,17 @@ function ($vv) use (&$inBindings, $k) {
                     $whereClauseSegments[] = "\"$k\" IN (:" . implode(', :', array_keys($inBindings)) . ")";
                     $bindings += $inBindings;
                 } else {
-                    $bindings["$k"] = $v;
-                    $whereClauseSegments[] = "\"$k\" = :{$k}";
+                    //Potential hack here, adding date checks would make this un-reusable but have stronger typing.
+                    if ($k == 'start_date') {
+                        $bindings["$k"] = $v;
```
Comment (samkeen) [inline src/Tools/FormExporter.php]: take out  "'s for "$k"

File: src/Tools/FormExporter.php [hunk 2]
```diff
@@ -294,4 +309,99 @@ function ($vv) use (&$inBindings, $k) {
...
+        // three: check that start date <= end date (if they are equal, we'll go 00:00:00 - 23:59:59)
+        // If any of these fail, throw an InvalidArgumentException("usable data");
+
+        // If we didn't get any criteria, we'll consider it validated for now.
+        if (is_array($selectCriteria) && count($selectCriteria) == 0) {
+            return true;
+        }
+
+        // We _could_ treat a solo start_date as a single-day option here.
+        if (isset($selectCriteria['start_date']) && !isset($selectCriteria['end_date'])) {
```
Comment (samkeen) [inline src/Tools/FormExporter.php]: extract method on these 2 if blocks

File: src/Tools/FormExporter.php [hunk 3]
```diff
@@ -294,4 +309,99 @@ function ($vv) use (&$inBindings, $k) {
...
+            throw new \InvalidArgumentException("start_date exists without an end_date.");
+        }
+
+        if (!isset($selectCriteria['start_date']) && isset($selectCriteria['end_date'])) {
+            throw new \InvalidArgumentException("end_date exists without a start_date.");
+        }
+
+        if (isset($selectCriteria['start_date']) && isset($selectCriteria['end_date'])) {
```
Comment (samkeen) [inline src/Tools/FormExporter.php]: extract method

File: src/Tools/FormExporter.php [hunk 4]
```diff
@@ -294,4 +309,99 @@ function ($vv) use (&$inBindings, $k) {
...
+        }
+
+        // Now for the actual data validation.
+        // Ignore this for now
+//        $x = \Igniter\Common\DateTimeUtil::dateTimeIsValid($startDate);
+//        $x;
+
+        // Check if the start date is in the future.
+        $todayDiff = $startDate->diff(new \DateTime());
+        if ($todayDiff->days > 0 && $todayDiff->invert == 1) {
```
Comment (samkeen) [inline src/Tools/FormExporter.php:371]: extract method

File: src/Tools/FormExporter.php [hunk 5]
```diff
@@ -294,4 +309,99 @@ function ($vv) use (&$inBindings, $k) {
...
+        return true;
+    }
+
+    /**
+     * @param string $startDate
+     * @return \DateTime
+     */
+    protected function getStartDate($startDate)
+    {
+        return new \DateTime($startDate);
```
Comment (samkeen) [inline src/Tools/FormExporter.php:399]: copy pasta check for warning

File: tests/Tools/ExporterTest.php [hunk 6]
```diff
@@ -389,5 +413,362 @@ function testUnWritableOutputDirThrowExpectedException()
...
+        $validated = TestHelper::invokeNonPublicMethod($exporter, 'validateSelectCriteria',
+            [['start_date' => '2015-02-01', 'end_date' => '2015-02-01']]);
+        $this->assertTrue($validated, "Start date == end date failed to validate");
+    }
+
+    /**
+     * @group FL-1161
+     * @group FL-1236
+     */
+    function testValidateSelectCriteriaStartDateInTheFuture()
```
Comment (samkeen) [inline tests/Tools/ExporterTest.php]: Phrasing!!

File: src/Tools/FormExporter.php [hunk 7]
```diff
@@ -294,4 +309,99 @@ function ($vv) use (&$inBindings, $k) {
...
+        if (is_array($selectCriteria) && count($selectCriteria) == 0) {
+            return true;
+        }
+
+        // We _could_ treat a solo start_date as a single-day option here.
+        if (isset($selectCriteria['start_date']) && !isset($selectCriteria['end_date'])) {
+            throw new \InvalidArgumentException("start_date exists without an end_date.");
+        }
+
+        if (!isset($selectCriteria['start_date']) && isset($selectCriteria['end_date'])) {
```
Comment (samkeen) [inline src/Tools/FormExporter.php]: we roll camelCase

File: src/Tools/FormExporter.php (hunk 1 above)
Comment (onewheelskyward) [inline src/Tools/FormExporter.php]: done

File: tests/Tools/ExporterTest.php (hunk 6 above)
Comment (onewheelskyward) [inline tests/Tools/ExporterTest.php]: Noooooooooooooooooooooooooooooooooooooooooooooooooooooooope!

File: src/Tools/FormExporter.php (hunk 7 above)
Comment (onewheelskyward) [inline src/Tools/FormExporter.php]: madeItSo

**Pull request closed without merging**